### PR TITLE
feat!: convert react-interpolate to TypeScript

### PR DESCRIPTION
**Pull request by gnapse**
This PR converts `react-interpolate` source and tests to TypeScript, replaces the handwritten declaration file with generated declarations, and updates the existing build/test tooling to support TS inputs.

It also tightens the parser/token/AST typing model and enables stricter TypeScript compiler checks, including isolated declaration emit.

The published CJS and ESM entry points are preserved, and runtime behavior was compared against `v2.2.2` for documented usage.

Validation: `npm run typecheck`, `npm run lint`, `npm test -- --runInBand`, and `npm run build`.

## Reference

- [Modernizing @doist/react-interpolate](https://twist.com/a/1585/ch/190200/t/7798940/)

### CODE REVIEW

**Review by doistbot — COMMENTED** [review]
This PR successfully converts `react-interpolate` to TypeScript, updating the build tooling and tightening the AST and parser typings while preserving the existing CJS and ESM entry points. This is a highly valuable upgrade that establishes a stricter, more reliable foundation for future development. A few areas have been flagged for further review, including managing Babel runtime dependencies, ensuring type compatibility for CommonJS consumers, narrowing overly broad types, and addressing a few minor performance, test isolation, and infinite loop edge cases.

<sub>[Share Feedback](https://github.com/Doist/doistbot-app/issues/new?template=review-feedback.yml&title=Review+feedback+for+Doist%2Freact-interpolate&pr_url=https%3A%2F%2Fgithub.com%2FDoist%2Freact-interpolate%2Fpull%2F55&datadog_url=https%3A%2F%2Fapp.datadoghq.com%2Flogs%3Fquery%3Dservice%253A%2528doistbot-app-sandbox%2520OR%2520doistbot-app%2529%2520%2540meta.deliveryId%253A4bd09120-553a-11f1-8fdd-f9009eca1459%26agg_m%3Dcount%26agg_m_source%3Dbase%26agg_t%3Dcount%26clustering_pattern_field_path%3Dmessage%26cols%3Dhost%252Cservice%26fromUser%3Dtrue%26messageDisplay%3Dinline%26refresh_mode%3Dsliding%26storage%3Dhot%26stream_sort%3Ddesc%26viz%3Dstream%26live%3Dtrue) • [Review Logs](https://app.datadoghq.com/logs?query=service%3A%28doistbot-app-sandbox%20OR%20doistbot-app%29%20%40meta.deliveryId%3A4bd09120-553a-11f1-8fdd-f9009eca1459&agg_m=count&agg_m_source=base&agg_t=count&clustering_pattern_field_path=message&cols=host%2Cservice&fromUser=true&messageDisplay=inline&refresh_mode=sliding&storage=hot&stream_sort=desc&viz=stream&live=true)</sub>

File: rollup.config.js [hunk 1]
```diff
@@ -1,23 +1,29 @@
...
+import { nodeResolve } from '@rollup/plugin-node-resolve'
 import pkg from './package.json'
-import copy from 'rollup-plugin-copy'
+
+const extensions = ['.js', '.jsx', '.ts', '.tsx']
 
 export default {
-    input: 'src/index.js',
+    input: 'src/index.ts',
+    external: [...Object.keys(pkg.peerDependencies || {}), /@babel\/runtime/],
```
Comment (doistbot) [inline rollup.config.js:9]: [P1] `@babel/runtime` is now explicitly externalized, but it is still only a devDependency. The built bundles will import helpers from `@babel/runtime/*`, so consumers of the published package can hit `Cannot find module '@babel/runtime/…'` at runtime. Either move `@babel/runtime` to `dependencies` or stop externalizing it so the helpers stay bundled.

File: package.json [hunk 2]
```diff
@@ -6,7 +6,7 @@
     "description": "A string interpolation component that formats and interpolates a template string in a safe way",
     "main": "dist/react-interpolate.cjs",
     "module": "dist/react-interpolate.mjs",
-    "types": "dist/react-interpolate.d.ts",
+    "types": "dist/index.d.ts",
```
Comment (doistbot) [inline package.json:9]: [P2] Switching `types` to the generated `dist/index.d.ts` changes the package’s declaration shape from the old CommonJS-style shim to ESM re-exports, while `main` still points at a CJS entrypoint. That drops type support for TS consumers using `require()` / `import = require()` against the CJS build. If the CJS entrypoint is meant to stay supported, keep a CJS-compatible declaration shim or add conditional `exports`/`types` for the two module formats.

File: src/interpolate.tsx [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+export type Mapping = Record<string, MappingValue>
+
+export interface InterpolateProps {
+    string: string
+    syntax?: Syntax
+    mapping?: Mapping
+    graceful?: boolean
+}
+
+type InterpolatePropTypes = {
```
Comment (doistbot) [inline src/interpolate.tsx]: [P3] The explicit `InterpolatePropTypes` type duplicates the shape of the `interpolatePropTypes` object below. You can remove this type definition entirely and let TypeScript infer it by using `typeof interpolatePropTypes` where needed.

File: src/lexer.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    matches.forEach((match) => {
+        if (match.start === start) {
+            start = match.end
+            return
+        }
+
+        const end = match.start
+        const text = string.substring(start, end)
+
+        textTokens.push({
```
Comment (doistbot) [inline src/lexer.ts]: [P3] The object literal structure for creating a text token is duplicated here and on line 105. Consider extracting a `createTextToken(text, start, end)` helper to share this logic and match the pattern established by `createSyntaxToken`.

File: src/lexer.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+export default function lexer(string: string, syntax: Syntax): Token[] {
+    const matches: SyntaxToken[] = []
+
+    for (const rule of syntax) {
+        const { type, regex } = rule
+
+        regex.lastIndex = 0
+
+        let result: RegExpExecArray | null
+        while ((result = regex.exec(string)) !== null) {
```
Comment (doistbot) [inline src/lexer.ts:88]: [P2] If a user provides a custom `Syntax` rule with a regex that lacks the global (`g`) flag, `regex.exec` will not advance `lastIndex`, causing an infinite loop. Consider adding a defensive check (e.g., `if (!regex.global) throw new Error(...)`) before the loop to prevent runtime hangs.

File: src/interpolate.tsx [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    NODE_TAG_ELEMENT,
+    NODE_TEXT,
+    NODE_VOID_ELEMENT,
+} from './constants'
+import type { SyntaxNode } from './node'
+import parser from './parser'
+import type { Syntax } from './syntax'
+
+export type MappingRenderFunction = (children?: React.ReactNode) => React.ReactNode
+export type MappingValue = React.ReactNode | MappingRenderFunction
```
Comment (doistbot) [inline src/interpolate.tsx:14]: [P2] `MappingValue` is broader than the runtime contract for tag mappings. `React.ReactNode` includes strings, numbers, booleans, etc., so `mapping={{ b: 'plain text' }}` now type-checks even though `NODE_TAG_ELEMENT` still throws unless the value is a render function or a valid React element. Please split placeholder/self-closing values from tag values, or otherwise narrow the exported type so the generated declarations stay sound.

File: src/parser.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        const token = this.match(TOKEN_OPEN_TAG)
+        this.pushTag(token)
+
+        return token
+    }
+
+    elementOrData(): SyntaxNode[] {
+        const children: SyntaxNode[] = []
+
+        while (this.predict(TOKEN_OPEN_TAG, TOKEN_SELF_TAG, TOKEN_TEXT, TOKEN_PLACEHOLDER)) {
```
Comment (doistbot) [inline src/parser.ts]: [P2] The `predict` method creates a new array via rest parameters on each call. Calling it repeatedly in this loop adds unnecessary memory allocation overhead per token. Replacing this chain with a `switch (this.lookahead.type)` block avoids these repeated allocations and simplifies the branching logic.

File: __test__/Interpolate.test.tsx [hunk 8]
```diff
@@ -1,30 +1,34 @@
...
-Interpolate.defaultProps = {
-    graceful: false,
-}
+import React from 'react'
+import Interpolate, { SYNTAX_I18NEXT, type InterpolateProps } from '../src'
 
-const surpressConsole = () => {
-    const w = jest.spyOn(console, 'warn').mockImplementation()
-    const e = jest.spyOn(console, 'error').mockImplementation()
+const suppressConsole = () => {
```
Comment (doistbot) [inline __test__/Interpolate.test.tsx]: [P2] Returning a manual cleanup function makes the tests brittle. If a test assertion fails before `restore()` is called, the console mock leaks to all subsequent tests, potentially hiding actual errors and making debugging difficult. Consider removing the manual restore and instead registering an `afterEach(() => jest.restoreAllMocks())` hook in this file, or wrap the test execution in a `try...finally` block.

File: __test__/Interpolate.test.tsx [hunk 9]
```diff
@@ -1,30 +1,34 @@
 /* eslint-disable react/display-name */
-import React from 'react'
 import { render } from '@testing-library/react'
-import Interpolate, { SYNTAX_I18NEXT } from '../src'
-
-Interpolate.defaultProps = {
-    graceful: false,
-}
+import React from 'react'
+import Interpolate, { SYNTAX_I18NEXT, type InterpolateProps } from '../src'
```
Comment (doistbot) [inline __test__/Interpolate.test.tsx:4]: [P2] These tests still import `../src`, so they won't catch regressions in the consumer-facing package surface changed by this PR (`dist` entry points and generated declarations). Please add at least one smoke/type test that imports the built package the way consumers do and compiles against the published typings.

File: src/interpolate.tsx (hunk 3 above)
Comment (gnapse) [reply src/interpolate.tsx]: Actually, I'll remove the usage of prop-types in this repo entirely.

File: package.json (hunk 2 above)
Comment (gnapse) [reply package.json:9]: Good catch on the package contract risk. I didn’t restore the old `export =` thing because that no longer matches the actual CJS runtime shape on this branch. Instead, I added explicit package-level compatibility checks for both ESM imports and CommonJS TypeScript consumption (`import pkg = require(...)` / `pkg.default`), so the generated declarations are now validated against the contract we actually ship.

File: rollup.config.js (hunk 1 above)
Comment (gnapse) [reply rollup.config.js:9]: Even though we'll me migrating off of rollup+babel soon, I still fixed this to keep this PR consistent internally.

**Comment by craigcarlyle** [comment]
🔥 ✅
